src/adapters/MorphoMarketV1Adapter.sol

```diff
@@ -95,21 +99,15 @@ contract MorphoMarketV1Adapter is IMorphoMarketV1Adapter {
         require(msg.sender == parentVault, NotAuthorized());
         require(marketParams.loanToken == asset, LoanAssetMismatch());
 
-        if (assets > 0) {
-            (, uint256 redeemedShares) = IMorpho(morpho).withdraw(marketParams, assets, 0, address(this), address(this));
-            shares[marketId] -= redeemedShares;
-        }
-        // TODO
-        int256 change = int256(expectedSupplyAssets(marketParams, shares[marketId])) - int256(allocation(marketParams));
-
-        if (shares[marketId] == 0 && assets > 0) {
-            for (uint256 i = 0; i < allMarketParams.length; i++) {
-                if (Id.unwrap(allMarketParams[i].id()) == Id.unwrap(marketId)) {
-                    allMarketParams[i] = allMarketParams[allMarketParams.length - 1];
-                    allMarketParams.pop();
-                    break;
-                }
-            }
+        if (assets > 0) IMorpho(morpho).withdraw(marketParams, assets, 0, address(this), address(this));
+        // TODO: justify safe casting
+        int256 change =
+            int256(IMorpho(morpho).expectedSupplyAssets(marketParams, address(this))) - int256(allocation(marketParams));
+
+        if (indexInListCtz[marketId] != 0 && IMorpho(morpho).supplyShares(marketId, address(this)) == 0) {
+            allMarketParams[ctz(indexInListCtz[marketId])] = allMarketParams[allMarketParams.length - 1];
```

Comment:
missing index update for indexInListCtz[allMarketParams[allMarketParams.length - 1]];